habitat-lab/habitat/tasks/rearrange/actions/actions.py

```diff
@@ -469,6 +469,178 @@ def step(self, *args, is_last_action, **kwargs):
+            goal_pos.append(
+                target_trans.transform_point(
+                    np.array([self._x_offset[i], 0, self._y_offset[i]])
+                )
+            )
+
+        # For step filter of offset positions
+        end_pos = []
+        for i in range(num_check_cylinder):
+            pos = self._sim.step_filter(cur_pos[i], goal_pos[i])
+            # Sanitize the height
+            pos[1] = cur_pos[i][1]
+            end_pos.append(pos)
+
+        # For step filter of the center original pos
+        center_end_pos[1] = trans.translation[1]
+
+        # Planar move distance clamped by NavMesh
+        move = []
+        for i in range(num_check_cylinder):
+            move.append((end_pos[i] - goal_pos[i]).length())
+        center_move = (center_end_pos - trans.translation).length()
+
+        # Wrap the move direction if we use sliding
+        # Find the largest diff moving direction, which means that there is a collision in that cylinder
+        max_idx = np.argmax(move + [center_move])
+        move_vec = (end_pos + [center_end_pos])[max_idx] - (
+            cur_pos + [trans.translation]
+        )[max_idx]
+        new_end_pos = trans.translation + move_vec
+        new_target_trans = mn.Matrix4.from_(
+            target_rigid_state.rotation.to_matrix(), new_end_pos
+        )
+
+        # For detection of linear or angualr velocities
+        # There is a collision if the difference between the clamped NavMesh position and target position is too great for any point.
+        lin_diff = len([v for v in move if v > self._lin_collision_threshold])
+        ang_diff = len([v for v in move if v > self._ang_collision_threshold])
+
+        if (lin_diff > 0) and center_move != 0:
+            return True, new_target_trans
+        elif (ang_diff > 0) and center_move == 0:
+            return True, new_target_trans
+        else:
+            return False, new_target_trans
+
+    def update_base(self):
+        ctrl_freq = self._sim.ctrl_freq
+
+        trans = self.cur_articulated_agent.sim_obj.transformation
+        rigid_state = habitat_sim.RigidState(
+            mn.Quaternion.from_matrix(trans.rotation()), trans.translation
+        )
+
+        target_rigid_state = self.base_vel_ctrl.integrate_transform(
+            1 / ctrl_freq, rigid_state
+        )
+        end_pos = self._sim.step_filter(
+            rigid_state.translation, target_rigid_state.translation
+        )
+
+        end_pos -= self.cur_articulated_agent.params.base_offset
+        target_trans = mn.Matrix4.from_(
+            target_rigid_state.rotation.to_matrix(), end_pos
+        )
+
+        did_coll, new_target_trans = self.collision_check(
+            trans, target_trans, end_pos, target_rigid_state
+        )
+        self.cur_articulated_agent.sim_obj.transformation = new_target_trans
+
+        if not self._allow_dyn_slide:
+            if did_coll:
+                self.cur_articulated_agent.sim_obj.transformation = trans
```

Comment:
We should not allow a step if there was collision and sliding is off.

Reply:
yes, this is exactly what I am doing here. Thank you!